File: ios/Classes/ThumbnailGenerator.swift
import AVFoundation
import UIKit

/// Extracts and caches JPEG thumbnail frames from video files.
///
/// Thumbnails are stored in `NSTemporaryDirectory()/vup_thumbs/` and keyed
/// by `{hash(videoPath)}_{timestampMs}_{width}.jpg`. A second call with the
/// same arguments is served from the cache without re-extraction.
final class ThumbnailGenerator {
    static let shared = ThumbnailGenerator()

    private let thumbDir: URL

    private init() {
        let tmp = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("vup_thumbs")
        try? FileManager.default.createDirectory(
            at: tmp,
            withIntermediateDirectories: true
        )
        thumbDir = tmp
    }

    /// Generates thumbnails for [videoPath] at each timestamp in [timestampsMs]
    /// (milliseconds) scaled to [width] pixels wide.
    ///
    /// Executes synchronously on a background thread and calls [completion]
    /// on that same background thread with the ordered list of cached file
    /// paths. Callers must dispatch back to the main thread when needed.
    ///
    /// Zero-tolerance extraction (`requestedTimeToleranceBefore/After = .zero`)
    /// gives frame-exact results at the cost of higher latency. Use a small
    /// non-zero tolerance (e.g. `CMTimeMake(1, 10)`) if speed is preferred
    /// over precision.
    func generate(
        videoPath: String,
        timestampsMs: [Int],
        width: Int,
        completion: @escaping ([String]) -> Void
    ) {
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self else {
                completion([])
                return
            }

            let url = URL(fileURLWithPath: videoPath)
            let asset = AVURLAsset(url: url)
            let generator = AVAssetImageGenerator(asset: asset)
            generator.appliesPreferredTrackTransform = true
            // Zero tolerance: frame-exact extraction (slower but precise).
            // Tradeoff: replace with CMTimeMake(1, 10) for better performance
            // when exact frame accuracy is not required.
            generator.requestedTimeToleranceBefore = .zero
            generator.requestedTimeToleranceAfter = .zero

            var paths: [String] = []
            for tsMs in timestampsMs {
                let cacheKey = "\(self.hashPath(videoPath))_\(tsMs)_\(width)"
                let fileURL = self.thumbDir
                    .appendingPathComponent("\(cacheKey).jpg")

                if FileManager.default.fileExists(atPath: fileURL.path) {
                    paths.append(fileURL.path)
                    continue
                }

                let time = CMTime(
                    value: CMTimeValue(tsMs),
                    timescale: 1000
                )
                guard let cgImage = try? generator.copyCGImage(
                    at: time,
                    actualTime: nil
                ) else {
                    continue
                }

                let uiImage = UIImage(cgImage: cgImage)
                let scaled = self.scale(uiImage, toWidth: width)
                if let data = scaled.jpegData(compressionQuality: 0.8) {
                    try? data.write(to: fileURL)
                    paths.append(fileURL.path)
                }
            }

            completion(paths)
        }
    }

    // MARK: - Private helpers

    /// djb2 hash of the path encoded as a hex string — used as the cache-key
    /// prefix to avoid path-traversal issues and long file names.
    private func hashPath(_ path: String) -> String {
        var hash: UInt64 = 5381
        for byte in path.utf8 {
            hash = hash &* 33 &+ UInt64(byte)
        }
        return String(hash, radix: 16)
    }

    private func scale(_ image: UIImage, toWidth width: Int) -> UIImage {
        let targetWidth = CGFloat(width)
        let scale = targetWidth / image.size.width
        let newSize = CGSize(
            width: targetWidth,
            height: (image.size.height * scale).rounded()
        )
        let renderer = UIGraphicsImageRenderer(size: newSize)
        return renderer.image { _ in
            image.draw(in: CGRect(origin: .zero, size: newSize))
        }
    }
}

File: android/src/main/kotlin/com/andre/video_ultra_player/ThumbnailGenerator.kt
package com.andre.video_ultra_player

import android.content.Context
import android.graphics.Bitmap
import android.media.MediaMetadataRetriever
import android.os.Build
import java.io.File
import java.io.FileOutputStream

/**
 * Extracts and caches JPEG thumbnail frames from video files.
 *
 * Thumbnails are stored in `context.cacheDir/vup_thumbs/` and keyed by
 * `{hash(videoPath)}_{timestampMs}_{width}.jpg`. A second call with the
 * same arguments is served from disk cache without re-extraction.
 */
internal class ThumbnailGenerator(private val context: Context) {

    private val thumbDir: File by lazy {
        File(context.cacheDir, "vup_thumbs").also { it.mkdirs() }
    }

    /**
     * Generates thumbnails for [videoPath] at each timestamp in [timestampsMs]
     * (milliseconds) scaled to [width] pixels wide (height is proportional).
     *
     * Must NOT be called on the main thread. Returns an ordered list of
     * absolute paths to the cached JPEG files. Timestamps for which extraction
     * fails are omitted from the result.
     */
    fun generate(videoPath: String, timestampsMs: List<Int>, width: Int): List<String> {
        val retriever = MediaMetadataRetriever()
        return try {
            retriever.setDataSource(videoPath)
            timestampsMs.mapNotNull { tsMs ->
                generateFrame(retriever, videoPath, tsMs, width)
            }
        } finally {
            retriever.release()
        }
    }

    // ── Private helpers ──────────────────────────────────────────────────────

    private fun generateFrame(
        retriever: MediaMetadataRetriever,
        videoPath: String,
        tsMs: Int,
        width: Int,
    ): String? {
        val cacheKey = "${hashPath(videoPath)}_${tsMs}_$width"
        val file = File(thumbDir, "$cacheKey.jpg")
        if (file.exists()) return file.absolutePath

        val timeUs = tsMs.toLong() * 1_000L

        val bitmap: Bitmap = (if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1) {
            val nativeW = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_VIDEO_WIDTH)?.toIntOrNull() ?: 1
            val nativeH = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_VIDEO_HEIGHT)?.toIntOrNull() ?: 1
            val scaledHeight = (nativeH.toLong() * width / nativeW).toInt().coerceAtLeast(1)
            retriever.getScaledFrameAtTime(
                timeUs,
                MediaMetadataRetriever.OPTION_CLOSEST_SYNC,
                width,
                scaledHeight,
            )
        } else {
            // API 24–26: retrieve the full frame then scale down.
            val full = retriever.getFrameAtTime(
                timeUs,
                MediaMetadataRetriever.OPTION_CLOSEST_SYNC,
            ) ?: return null
            val scaledHeight = (full.height.toLong() * width / full.width).toInt()
                .coerceAtLeast(1)
            Bitmap.createScaledBitmap(full, width, scaledHeight, /* filter = */ true)
        }) ?: return null

        return try {
            FileOutputStream(file).use { out ->
                bitmap.compress(Bitmap.CompressFormat.JPEG, 80, out)
            }
            file.absolutePath
        } catch (_: Throwable) {
            null
        }
    }

    /**
     * djb2 hash of [path] encoded as a hex string — used as the cache-key
     * prefix to avoid path-traversal issues and overly long file names.
     */
    private fun hashPath(path: String): String {
        var hash = 5381L
        for (c in path) {
            hash = hash * 33 + c.code
        }
        return hash.toString(16)
    }
}
